### PR TITLE
Fix chat-chain pipelined splits: verify-history accounting, early-stop replies, session leaks

### PR DESCRIPTION
**Pull request by danielwinterw**
# Fix chat-chain pipelined splits: verify-history accounting, early-stop replies, session leaks

Pipelined verify windows (`verify_window_pipeline_depth >= 2`) with chat
sampling are currently broken on main for any staged split: requests fail with
`sampled verification history is behind sampler state` (reliably at depth 3,
intermittently at depth 2), and windows that do pass silently corrupt the chat
sampler's absorbed history. Bisected + instrumented on a 2-process loopback
Qwen3-8B split (standalone suffix, temperature 0).

## What broke

`execution-batch.cpp` only eagerly records decoded tokens into
`token_history` when the batch has no activation input:

```cpp
session->n_past += 1;
if (!activation_input) {
    skippy_record_tokens(session, &token_ids[i], 1);
}
```

but `verification.cpp`'s `include_output` block unconditionally removes that
eager append (`history.size() - token_count` + truncate) before re-recording
rows as they are sampled. On any stage fed by an upstream stage
(`skippy_verify_activation_frame`), the append never happened, so:

1. each passing window drops real history rows and re-records sampled rows at
   wrong logical positions — the chat sampler absorbs corrupted history;
2. once `sampling_accepted_token_count` catches up to `token_history.size()`,
   the next contiguous window trips the invariant and the request dies.

Instrumented trip: `n_past=372 history=339 accepted=339 token_count=33` —
n_past advanced by the batch, history did not.

## Fixes

- **llama patch 0026**: in the `include_output` block, only subtract/truncate
  for token-batch verification; activation-frame verification's history
  already ends at the pre-batch position.
- **Early-stopped verify replies**: stateful chat sampling stops at the first
  mismatched row and returns only the sampled rows; the pipelined composer
  treated any short prediction array as a protocol error. Pass through the
  available rows — window classification already rejects at the divergence.
- **Prefix-cache session leaks**: two partial-restore early-outs
  (full-prompt-first-token chain restore, local prefill restore) returned
  without dropping the session, so identical-prompt retries exhausted the
  execution lanes and cascaded into 502s. Same fix shape as the existing
  `drop_embedded_split_restore` call sites.
- **`SKIPPY_KV_CACHE=off`** now overrides the planned stage kv-cache config
  (explicit disabled value only), so benches/incident response can turn the
  prefix cache off without replanning a topology.

## Validation

- 2-process loopback split (Qwen3-8B Q4_K_M, runtime-slice 0..18/18..36,
  embedded stage-0 OpenAI, standalone suffix 5/32/48, window 32, temp 0):
  previously every request failed at depth >= 2; with these fixes depth 2/3
  run to completion with exact outputs (accept 0.89, ~90-100 tok/s loopback).
- skippy-server suite, mesh-llm-config, skippy-protocol, and the
  mesh-llm-host-runtime suite green; clippy clean; `check-release` and
  crate-list consistency pass.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added support for configuring suffix-based speculative decoding with n-gram mode.

* **Bug Fixes**
  * Improved handling of empty and partial prediction results during generation.
  * Prevented stale execution sessions after incomplete prefix-cache restores.
  * Ensured caching is disabled when either cache setting is turned off.
  * Corrected token verification history tracking for more reliable decoding.

* **Tests**
  * Added coverage for n-gram decoding, prediction edge cases, and cache configuration combinations.

### CODE REVIEW

**Comment by michaelneale** [comment]
Thanks. Taking a look.

File: crates/mesh-llm-host-runtime/src/inference/skippy/resolver/tests.rs [hunk 1]
```diff
@@ -1494,3 +1494,88 @@ fn integrated_invalid_fixture_fails_closed_for_request_defaults_and_single_stage
...
+        None,
+    )
+    .expect("explicit ngram-suffix must resolve");
+    let decode = resolved.decode;
+    assert_eq!(decode.effective_strategy, "ngram-suffix", "effective strategy");
+    let ngram = decode.ngram.expect("suffix proposer must be present");
+    assert_eq!(ngram.min_ngram, 5);
+    assert_eq!(ngram.max_ngram, 32);
+    assert_eq!(decode.verify_window.pipeline_depth, 2);
+    assert_eq!(resolved.mode, "ngram", "resolved mode drives the embedded frontend");
```
[the remark this answers is not in the document]
Comment (danielwinterw) [reply crates/mesh-llm-host-runtime/src/inference/skippy/resolver/tests.rs]: Added in a367fe2: both tests now assert the suffix kind and max_proposal_tokens at the resolved and translated boundaries.

File: crates/skippy-server/src/frontend/embedded_generation/lifecycle.rs [hunk 2]
```diff
@@ -89,28 +89,28 @@ pub(super) fn compose_target_predictions(
...
-            "continuation verify window returned {} predictions; expected {proposal_count}",
-            traversal_predictions.len()
-        )));
-    }
-    let mut predictions = Vec::with_capacity(required);
+    let available = proposal_count.min(traversal_predictions.len());
+    let mut predictions = Vec::with_capacity(available.saturating_add(1));
     predictions.push(boundary);
-    predictions.extend_from_slice(&traversal_predictions[..proposal_count]);
+    predictions.extend_from_slice(&traversal_predictions[..available]);
```
[the remark this answers is not in the document]
Comment (danielwinterw) [reply crates/skippy-server/src/frontend/embedded_generation/lifecycle.rs:118]: Fixed in a367fe2: empty continuation replies are now a backend error, matching the epoch-start path; added tests for the early-stop pass-through and both empty cases.

File: crates/skippy-server/src/kv_integration/config.rs [hunk 3]
```diff
@@ -242,6 +242,15 @@ fn infer_cache_payload(config: &StageConfig) -> StagePrefixCachePayload {
 }
 
 fn effective_cache_config(config: &StageConfig) -> Option<StageKvCacheConfig> {
+    // An explicit environment kill-switch beats the planned stage config so
+    // benches and incident response can turn the prefix cache off without
+    // replanning the topology.
+    if let Ok(value) =
+        std::env::var("SKIPPY_KV_CACHE").or_else(|_| std::env::var("SKIPPY_PREFIX_CACHE"))
+        && parse_cache_mode(&value) == Some(StageKvCacheMode::Disabled)
```
[the remark this answers is not in the document]
Comment (danielwinterw) [reply crates/skippy-server/src/kv_integration/config.rs]: Fixed in a367fe2: both variables are evaluated independently via cache_disabled_by_env, with a unit test covering the on/off combinations.

**Review by i386 — APPROVED** [review]
Reviewed the activation-input history accounting, early-stop prediction composition, prefix-restore cleanup, cache kill-switch precedence, and regression coverage. The llama.cpp patch is behavior-only (no ABI surface change), the patch/static-ABI and full PR checks are green, and the previously raised inline findings are resolved. I found no remaining blocking issues.

**Review by michaelneale — APPROVED** [review]

**Comment by michaelneale** [comment]
Housekeeping on the patch queue, non-blocking — flagging before merge so it doesn't get absorbed silently. cc @i386.

**This is a rebase collision, not an authoring error.** Timeline in UTC:

| | |
|---|---|
| this branch's head `7ae8bde1a` | 2026-08-22T10:45:41Z |
| #1402 lands on `main` (`331852356`) | 2026-08-22T20:47:31Z |

At the time `0026` was written the queue ended at `0025-skippy-prune-unverified-MTP-proposal-rows-from-the-s.patch`, so `0026` was the correct next number. #1402 renumbered the queue to `0001-0015` ten hours later. Nothing to fix in how this was done.

**Effect after merge.** `git merge-tree origin/main <this branch>` gives a queue of `0001`…`0015` + `0026` — a ten-slot hole. `.agents/skills/llama-patch-changes/SKILL.md:25` asks for numbers to be "unique and contiguous".

**No gate catches this.** `scripts/prepare-llama.sh:138` globs `*.patch` and `sort`s lexically, so `0026` applies last either way and the build is unaffected. `compute_patch_digest` (`:181-192`) hashes relative filename + contents, so renaming the file *does* move the digest and invalidate cached prepared checkouts — worth doing in one step rather than twice.

**Applies cleanly regardless.** I extracted `verification.cpp` from `main`'s `0008` and applied this patch against it standalone — clean, no fuzz. The hunk context (`history_size_before_verification`, `sampled verification token history underflow`) is untouched by #1402, which only renumbered and did not alter that region. So this is cosmetic only.

**Suggestion:** rename to `0016-skippy-fix-verify-history-accounting-for-activation-.patch` before merge. One-line change, no content edit. If the preference is to let it ride and renumber on the next queue rewrite, that's fine too — but then it's worth a line in the PR body so the next person to run `just diff` doesn't go looking for `0016`–`0025`.

One related note for whoever picks the follow-up: there's no CI check asserting queue contiguity, and #1402 + this PR is the second time in a week the numbering has drifted. A five-line test over `third_party/llama.cpp/patches/` would make it self-enforcing.

Read at `origin/main 8c3db3dbc` and this PR at `7ae8bde1a`.
